iavl/iavl_range_proofs.go

```diff
@@ -0,0 +1,456 @@
+
+// KeyRangeProof is proof that a range of keys does or does not exist.
+type KeyRangeProof struct {
+	RootHash   data.Bytes   `json:"root_hash"`
+	PathToKeys []*PathToKey `json:"paths"`
+
+	Left  *PathWithNode `json:"left"`
+	Right *PathWithNode `json:"right"`
+}
+
+// Verify that a range proof is valid.
+func (proof *KeyRangeProof) Verify(
+	startKey, endKey []byte, limit int, keys, values [][]byte, root []byte,
+) error {
+	if len(proof.PathToKeys) != len(keys) || len(values) != len(keys) {
+		return errors.New("wrong number of keys or values for proof")
+	}
+	if len(proof.PathToKeys) == 0 && proof.Left == nil && proof.Right == nil {
+		return errors.New("proof is incomplete")
+	}
+
+	// If startKey > endKey, reverse the keys and values, since our proofs are
+	// always in ascending order.
+	startKey, endKey, keys, values = reverseIfDescending(startKey, endKey, keys, values)
+
+	// Verify that all paths are adjacent to one another.
+	if err := proof.verifyPathAdjacency(); err != nil {
+		return err
+	}
+
+	firstKey, lastKey := startKey, endKey
+	if len(keys) > 0 {
+		firstKey, lastKey = keys[0], keys[len(keys)-1]
+	}
+	if err := verifyPaths(proof.Left, proof.Right, firstKey, lastKey, root); err != nil {
+		return err
+	}
+
+	// If proof.PathToKeys is empty, it means we have an empty range. This range
+	// can be between keys, or outside of the range of existing keys.
+	if len(keys) == 0 {
+		// Range is outisde and to the left of existing keys.
+		if proof.Left == nil && proof.Right.Path.isLeftmost() {
+			return nil
+		}
+		// Range is outisde and to the right of existing keys.
+		if proof.Right == nil && proof.Left.Path.isRightmost() {
+			return nil
+		}
+		// Range is between two existing keys.
+		if proof.Left != nil && proof.Right != nil {
+			return nil
+		}
+		return errors.New("invalid proof of empty range")
+	}
+
+	// If we've reached this point, it means our range isn't empty, and we have
+	// a list of keys.
+	for i, path := range proof.PathToKeys {
+		leafNode := IAVLProofLeafNode{KeyBytes: keys[i], ValueBytes: values[i]}
+		if err := path.verify(leafNode, root); err != nil {
+			return errors.Wrap(err, "failed to verify inner path")
+		}
+	}
+
+	if !bytes.Equal(startKey, keys[0]) && proof.Left == nil {
+		if !proof.PathToKeys[0].isLeftmost() {
+			return errors.New("left path is nil and first inner path is not leftmost")
+		}
+	}
+
+	// If the right proof is nil and the limit wasn't reached, we have to verify that
+	// we're not missing any keys. Basically, if a key to the right is missing because
+	// we've reached the limit, then it's fine. But if the key count is smaller than
+	// the limit, we need a right proof to make sure no keys are missing.
+	if proof.Right == nil && len(keys) != limit {
+		if !bytes.Equal(endKey, lastKey) && !proof.PathToKeys[len(proof.PathToKeys)-1].isRightmost() {
+			return errors.New("right path is nil and last inner path is not rightmost")
+		}
+	}
```

Comment:
So this means the user must pass in a limit equal to the length of the keys if they didnt set a particular limit in the query. Can we document that better on the method? 

Also we should probably check that len(keys) < limit. 

Otherwise I think you're right.

Do we have a spec for all this stuff? It would help to have a clear specification in one place that gave us somewhere to think through all the conditions.